Introducing 'dependency-tree.yaml' & 'component.yaml' files and combining all Component configs at the App build time

this opened the pull request:
- Introducing,
  -  'dependency-tree.yaml' which replaces raw 'dependency.tree' file in an App (issue #30)
  - 'component.yaml' which replaces 'bindings.yaml' in a Component
- Combine all the 'config.yaml' files of dependency Components into a single 'config.yaml' file when building an App (issue #23)
- Also fixes validation related issues #24, #25, #26, #27, #28, and #29
- Also,
  - Added missing javadocs.
  - Removed Maven Assembly Plugin API usage and dependency.
  - Added test cases for `DependencyTreeParser`.

Niranjan-K commented:
can we have a meaning full name in the PR? AFAIK we should not put PR comments as fixed with issue #

this commented:
@Niranjan-K Issue #23 and #30 are related. Issue #24, #25, #26, #27, #28 and #29 were also fixed in this PR as it was easier to fix them with the same changes. As your suggestion, renamed the title of this PR

chamilad commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/ComponentMojo.java:
```diff
@@ -24,43 +24,87 @@
+        // Validation: Parse component manifest file to make sure it is valid.
+        String componentManifestFilePath = pathOf(sourceDirectoryPath, FILE_COMPONENT_MANIFEST);
+        try {
+            new ComponentManifestParser().parse(componentManifestFilePath);
```
How frequent is this operation? At a glance this looks like we are creating a new object and discarding it for the GC to collect later.

this replied:
Changing the `ComponentManifestParser.parse` method to static will solve this.

chamilad commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/model/DependencyNode.java:
```diff
@@ -0,0 +1,172 @@
+            correctedArtifactId = artifactId.substring(0, (artifactId.length() - ARTIFACT_ID_TAIL_UI.length()));
+        } else {
+            correctedArtifactId = artifactId;
+        }
```
A blank line after a block ends helps code readability.

this replied:
fixed in commit 2a09a1892c422a1cc20b75995ac4094590fdcf59

Niranjan-K commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/ComponentManifestParser.java:
```diff
@@ -0,0 +1,69 @@
+        try {
+            return parseString(content);
+        } catch (Exception e) {
+            throw new ParsingException("Cannot parse the content of component manifest file '" + manifestFile + "'.",
```
have a single try and catch the two exceptions

vidurananayakkara replied:
Add this catch block to line number 56. Furthermore, it is rather better to use a specific exception type than "Exception"

this replied:
fixed in commit 6ea3ef1353558580a76bf0343c6df78a95f34497

vidurananayakkara commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/ComponentManifestParser.java:
```diff
@@ -0,0 +1,69 @@
+    public ComponentManifest parse(String componentManifestFilePath) throws ParsingException {
+        Path manifestFile = Paths.get(componentManifestFilePath);
+        if (!Files.exists(manifestFile)) {
+            return null; // File does not exists.
```
Throw custom exception here rather than returning null.

this replied:
`component.yaml` is optional for a Component, hence we shouldn't throw an exception here.

Kishanthan commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/AppMojo.java:
```diff
@@ -135,139 +112,277 @@
         @SuppressWarnings("unchecked")
         Set<Artifact> allDependencies = ((Set<Artifact>) project.getArtifacts());
-        Set<Artifact> uufComponentDependencies = allDependencies.stream()
+        Set<Artifact> allComponentDependencies = allDependencies.stream()
```
Can we see whether we could use Collections.groupBy here?

Kishanthan commented:
Need to look into how we can use the new maven 3.x API to get the dependency tree from Maven itself, rather than creating the dependency-tree file and them parsing again the to create our tree.

Niranjan-K commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/DependencyTreeParser.java:
```diff
@@ -0,0 +1,135 @@
+                currentParentNode = previousNode;
+                currentNode = createDependencyNode(line, currentParentNode);
+                currentParentNode.addDependency(currentNode);
+            } else {
```
It would be better to have debug logs here for each condition

this replied:
We should add logs for all Mojos. Created issue #32 for this.

Niranjan-K commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/DependencyTreeParser.java:
```diff
@@ -0,0 +1,135 @@
+     * @return root node of the dependency tree
+     * @throws ParsingException if an error occurred when parsing the dependency tree
+     */
+    DependencyNode parse(List<String> dependencyTreeLines) throws ParsingException {
```
As suggested by @afkham during the review, can you check how the dependencies can be retrieved using maven api.

this replied:
Let's add this as a future enhancement.

this commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/ThemeConfigParser.java:
```diff
@@ -0,0 +1,68 @@
+    public ThemeConfig parse(String themeConfigFilePath) throws ParsingException {
+        Path manifestFile = Paths.get(themeConfigFilePath);
+        if (!Files.exists(manifestFile)) {
+            return null; // File does not exists.
```
Instead of returning `null`, throw an exception because `theme.yaml` is mandatory for a Theme.

this replied:
fixed in commit 785cb1e595a4037830451a87134219ff52a62aec

this commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/ThemeConfigParser.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class ThemeConfigParser {
+
+    private final Yaml yaml = new Yaml();
```
fixed in commit 07e8ede1ba3cee1025213f6132306a7631d4407d

this commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/parser/ThemeConfigParser.java:
```diff
@@ -0,0 +1,68 @@
+            return null; // File does not exists.
+        }
+        String content;
+        try {
```
fixed in commit 07e8ede1ba3cee1025213f6132306a7631d4407d

this commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/util/ZipCreator.java:
```diff
@@ -0,0 +1,83 @@
+                                     String outputDirectoryPath,
+                                     String archiveFileName) throws MojoExecutionException {
+        String correctedBaseDirectory;
+        if (baseDirectoryName == null) {
```
fixed in commit f120ce4cdb743e6f1d2243d3ab2ced4e5af3e079

afkham commented:
Why is your ID @this? We need to be able to easily trace the github ID to
the real person so we have asked everybody to use their real names as much
as possible.

On Wed, Nov 30, 2016 at 10:18 PM, Sajith Janaprasad Ariyarathna <
notifications@github.com> wrote:

> *@this* commented on this pull request.
> ------------------------------
>
> In plugin/src/main/java/org/wso2/carbon/uuf/maven/ComponentMojo.java
> <https://github.com/wso2/carbon-uuf-maven-tools/pull/31>:
>
> > +            throw new MojoExecutionException(
> +                    "Packaging type of an UUF Component should be '" + ARTIFACT_TYPE_UUF_COMPONENT +
> +                            "'. Instead found '" + packaging + "'.");
> +        }
> +        // Validation: Parse configuration file to make sure it is a valid YAML file.
> +        String configFilePath = pathOf(sourceDirectoryPath, FILE_CONFIG);
> +        try {
> +            new ConfigurationParser().parse(configFilePath);
> +        } catch (ParsingException e) {
> +            throw new MojoExecutionException(
> +                    "Configuration file '" + configFilePath + "' of this UUF Component is invalid.", e);
> +        }
> +        // Validation: Parse component manifest file to make sure it is valid.
> +        String componentManifestFilePath = pathOf(sourceDirectoryPath, FILE_COMPONENT_MANIFEST);
> +        try {
> +            new ComponentManifestParser().parse(componentManifestFilePath);
>
> Changing the ComponentManifestParser.parse method to static will solve
> this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wso2/carbon-uuf-maven-tools/pull/31>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ACiBtWKANxp1Rhmsuu3AP2lh1tJaJW-mks5rDajYgaJpZM4K7Cdq>
> .
>



-- 
*Afkham Azeez*
Director of Architecture; WSO2, Inc.; http://wso2.com
Member; Apache Software Foundation; http://www.apache.org/
* <http://www.apache.org/>*
*email: **azeez@wso2.com* <azeez@wso2.com>
* cell: +94 77 3320919blog: **http://blog.afkham.org*
<http://blog.afkham.org>
*twitter: **http://twitter.com/afkham_azeez*
<http://twitter.com/afkham_azeez>
*linked-in: **http://lk.linkedin.com/in/afkhamazeez
<http://lk.linkedin.com/in/afkhamazeez>*

*Lean . Enterprise . Middleware*

imesh commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/AbstractUUFMojo.java:
```diff
@@ -0,0 +1,91 @@
+    protected static final String ARTIFACT_TYPE_UUF_THEME = "uuf-theme";
+
+    /**
+     * This Maven project.
```
It might be better to explain what we mean by "This Maven project".

imesh commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/AbstractUUFMojo.java:
```diff
@@ -0,0 +1,91 @@
+    protected String finalName;
+
+    /**
+     * Maven ProjectHelper.
```
Should "ProjectHelper" be written as "project helper"? Or use proper class reference:
http://docs.oracle.com/javase/1.5.0/docs/tooldocs/windows/javadoc.html#@see

imesh commented on plugin/src/main/java/org/wso2/carbon/uuf/maven/ComponentMojo.java:
```diff
@@ -24,43 +24,87 @@
+        // Validation: Packaging type should be 'uuf-component'.
+        if (!ARTIFACT_TYPE_UUF_COMPONENT.equals(packaging)) {
+            throw new MojoExecutionException(
+                    "Packaging type of an UUF Component should be '" + ARTIFACT_TYPE_UUF_COMPONENT +
```
The first letter of the word "Component" might not need to be in uppercase: "UUF component" might be much appropriate. Thanks!
